### ISSUE TITLE
[Inject] method called twice when registering a monoBehaviour present in the scene.

### ISSUE THREAD

**Issue by NeistH2o** [hadashiA/VContainer#230]
I have an internal `ClientInput` class implementing `MonoBehaviour` in my client scene, present in the hierarchy on startup.

The way I pass down its dependencies is via the following method:
```
[Inject]
internal void Construct(InputCaptureSettings settings)
{ ... }
```

The root `LifetimeScope` registers the instance present in the scene from a serialized field like so:
```
builder.RegisterComponent(_clientInput)
  .As<IClientInput>();
```

Just to understand how and when the dependencies are provided, I added a `Debug.Log` to the `Construct` method of `ClientInput` and it turns out that the method is called twice.

`ClientInput` isn't listed in the "auto inject gameobject" list of any scope.

Is there something I'm not doing correctly here or is it an unexpected behavior?

**Comment by hadashiA** [comment]
Hello, Thanks for the report.

I've tried RegisterComponent in a simple scene, but it doesn't seem to reproduce. `[Inject]` was called only once.

Is it possible to share the stack trace of two method calls?
Or could you share the code to reproduce it?

**Comment by NeistH2o** [comment]
Hi, thanks for the reply, I actually changed the code to not use any MonoBehaviour from the scene but I just created a new one to see if I could reproduce it:
```
using System.Collections.Generic;
using System.Collections;
using UnityEngine;
using VContainer;
using System;


namespace RAS.H2o
{
	internal interface IInjectedMonobehaviour : IDisposable { }

	internal class InjectedMonobehaviour : MonoBehaviour, IInjectedMonobehaviour
	{
		[Inject]
		internal void Construct()
		{
			Debug.Log("InjectedMonobehaviour Construct");
		}

		public void Dispose()
		{
			Debug.Log("Dispose");
		}
	}
}
```

Just like before I use a serialized field to pass it's reference down to the LifetimeScope (MonoBehaviour) and register it with `RegisterComponent(_injectedMonoBehaviour).As<IInjectedMonobehaviour>();`

The `[Inject]` method is again being called twice on startup, the first stack trace:
```
InjectedMonobehaviour Construct
UnityEngine.Debug:Log (object)
RAS.H2o.InjectedMonobehaviour:Construct () (at Assets/H2o/Head/client/Runtime/InjectedMonobehaviour.cs:17)
System.Reflection.MethodBase:Invoke (object,object[])
VContainer.Internal.ReflectionInjector:InjectMethods (object,VContainer.IObjectResolver,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/ReflectionInjector.cs:109)
VContainer.Internal.ReflectionInjector:Inject (object,VContainer.IObjectResolver,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/ReflectionInjector.cs:23)
VContainer.Unity.InstanceComponentRegistration:SpawnInstance (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/InstanceComponentRegistration.cs:32)
VContainer.Unity.ContainerBuilderUnityExtensions/<>c__5`1<RAS.H2o.InjectedMonobehaviour>:<RegisterComponent>b__5_0 (VContainer.IRegistration,VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/ContainerBuilderUnityExtensions.cs:113)
VContainer.ContainerBuilder:EmitCallbacks (VContainer.IObjectResolver,System.Collections.Generic.IEnumerable`1<System.ValueTuple`2<VContainer.IRegistration, System.Action`2<VContainer.IRegistration, VContainer.IObjectResolver>>>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/ContainerBuilder.cs:136)
VContainer.ContainerBuilder:Build () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/ContainerBuilder.cs:79)
VContainer.Unity.LifetimeScope:Build () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/LifetimeScope.cs:171)
VContainer.Unity.LifetimeScope:Awake () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/LifetimeScope.cs:130)
```

And second stack trace (longer):
```
InjectedMonobehaviour Construct
UnityEngine.Debug:Log (object)
RAS.H2o.InjectedMonobehaviour:Construct () (at Assets/H2o/Head/client/Runtime/InjectedMonobehaviour.cs:17)
System.Reflection.MethodBase:Invoke (object,object[])
VContainer.Internal.ReflectionInjector:InjectMethods (object,VContainer.IObjectResolver,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/ReflectionInjector.cs:109)
VContainer.Internal.ReflectionInjector:Inject (object,VContainer.IObjectResolver,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/ReflectionInjector.cs:23)
VContainer.Unity.InstanceComponentRegistration:SpawnInstance (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/InstanceComponentRegistration.cs:32)
VContainer.Container/<>c__DisplayClass5_0:<.ctor>b__1 () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:146)
System.Lazy`1<object>:get_Value ()
VContainer.Container:Resolve (VContainer.IRegistration) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:168)
VContainer.Container:Resolve (System.Type) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:154)
VContainer.IObjectResolverExtensions:ResolveOrParameter (VContainer.IObjectResolver,System.Type,string,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/IObjectResolverExtensions.cs:30)
VContainer.Internal.ReflectionInjector:CreateInstance (VContainer.IObjectResolver,System.Collections.Generic.IReadOnlyList`1<VContainer.IInjectParameter>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/ReflectionInjector.cs:39)
VContainer.Internal.Registration:SpawnInstance (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/Registration.cs:34)
VContainer.Container/<>c__DisplayClass5_0:<.ctor>b__1 () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:146)
System.Lazy`1<object>:get_Value ()
VContainer.Container:Resolve (VContainer.IRegistration) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:168)
VContainer.ScopedContainer:Resolve (VContainer.IRegistration) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:69)
VContainer.Internal.CollectionRegistration:SpawnInstance (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Internal/Registration.cs:87)
VContainer.ScopedContainer:Resolve (VContainer.IRegistration) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:65)
VContainer.ScopedContainer:Resolve (System.Type) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Container.cs:57)
VContainer.IObjectResolverExtensions:Resolve<System.Collections.Generic.IReadOnlyList`1<VContainer.Unity.IInitializable>> (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/IObjectResolverExtensions.cs:8)
VContainer.Unity.EntryPointDispatcher:Dispatch () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/EntryPointDispatcher.cs:34)
VContainer.Unity.EntryPointsBuilder/<>c:<EnsureDispatcherRegistered>b__0_0 (VContainer.IObjectResolver) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/ContainerBuilderUnityExtensions.cs:20)
VContainer.ContainerBuilder:EmitCallbacks (VContainer.IObjectResolver,System.Collections.Generic.IEnumerable`1<System.ValueTuple`2<VContainer.IRegistration, System.Action`2<VContainer.IRegistration, VContainer.IObjectResolver>>>) (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/ContainerBuilder.cs:143)
VContainer.ContainerBuilder:Build () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/ContainerBuilder.cs:79)
VContainer.Unity.LifetimeScope:Build () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/LifetimeScope.cs:171)
VContainer.Unity.LifetimeScope:Awake () (at Library/PackageCache/jp.hadashikick.vcontainer@1.8.0/Runtime/Unity/LifetimeScope.cs:130)
```

Hope this help, let me know if I should create a separate project and try to reproduce it there but my available time is very limited these days :s

**Comment by hadashiA** [comment]
Oops, My confirmation was insufficient.

There was a problem that when there was a Resolve from another class, it was called twice in total.

I've fixed this in #232.

**Comment by hadashiA** [comment]
Released in [1.8.1](https://github.com/hadashiA/VContainer/releases/tag/1.8.1)
